client/cli/src/logging.rs

```diff
@@ -0,0 +1,336 @@
+		FmtContext, FormatEvent, FormatFields,
+	},
+	layer::Context,
+	registry::LookupSpan,
+	Layer,
+};
+
+pub(crate) struct EventFormat<T = SystemTime> {
+	pub(crate) timer: T,
+	pub(crate) ansi: bool,
+	pub(crate) display_target: bool,
+	pub(crate) display_level: bool,
+	pub(crate) display_thread_name: bool,
+}
+
+// NOTE: the following code has been inspiref from tracing-subscriber
+//
+//       https://github.com/tokio-rs/tracing/blob/2f59b32/tracing-subscriber/src/fmt/format/mod.rs#L449
+impl<S, N, T> FormatEvent<S, N> for EventFormat<T>
+where
+	S: Subscriber + for<'a> LookupSpan<'a>,
+	N: for<'a> FormatFields<'a> + 'static,
+	T: FormatTime,
+{
+	fn format_event(
+		&self,
+		ctx: &FmtContext<S, N>,
+		writer: &mut dyn fmt::Write,
+		event: &Event,
+	) -> fmt::Result {
+		let normalized_meta = event.normalized_metadata();
+		let meta = normalized_meta.as_ref().unwrap_or_else(|| event.metadata());
+		time::write(&self.timer, writer, self.ansi)?;
+
+		if self.display_level {
+			let fmt_level = { FmtLevel::new(meta.level(), self.ansi) };
+			write!(writer, "{} ", fmt_level)?;
+		}
+
+		if self.display_thread_name {
+			let current_thread = std::thread::current();
+			match current_thread.name() {
+				Some(name) => {
+					write!(writer, "{} ", FmtThreadName::new(name))?;
+				}
+				// fall-back to thread id when name is absent and ids are not enabled
+				None => {
+					write!(writer, "{:0>2?} ", current_thread.id())?;
+				}
+			}
+		}
+
+		// Custom code to display node name
+		if let Some(span) = ctx.lookup_current() {
+			let parents = span.parents();
+			for span in std::iter::once(span).chain(parents) {
+				let exts = span.extensions();
+				if let Some(node_name) = exts.get::<NodeName>() {
+					write!(writer, "{}", node_name.as_str())?;
+					break;
+				}
+			}
+		}
+
+		let fmt_ctx = { FmtCtx::new(&ctx, event.parent(), self.ansi) };
+		write!(writer, "{}", fmt_ctx)?;
+		if self.display_target {
+			write!(writer, "{}:", meta.target())?;
+		}
+		ctx.format_fields(writer, event)?;
+		writeln!(writer)
+	}
+}
+
+pub(crate) struct NodeNameLayer;
+
+impl<S> Layer<S> for NodeNameLayer
+where
+	S: Subscriber + for<'a> LookupSpan<'a>,
+{
+	fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+		let span = ctx
+			.span(id)
+			.expect("new_span has been called for this span; qed");
+
+		if span.name() != "substrate-node" {
+			return;
+		}
+
+		let mut extensions = span.extensions_mut();
+
+		if extensions.get_mut::<NodeName>().is_none() {
+			let mut s = String::new();
+			let mut v = NodeNameVisitor(&mut s);
+			attrs.record(&mut v);
+
+			if !s.is_empty() {
+				let fmt_fields = NodeName(s);
+				extensions.insert(fmt_fields);
+			}
+		}
+	}
+}
+
+struct NodeNameVisitor<'a, W: std::fmt::Write>(&'a mut W);
+
+macro_rules! write_node_name {
+	($method:ident, $type:ty, $format:expr) => {
+		fn $method(&mut self, field: &tracing::field::Field, value: $type) {
+			if field.name() == "name" {
```

Comment:
Is there a chance of collision with the identifier `name` if someone else decides to use this in a span, or in practice is this impossible due to the order that the span hierarchy is traversed? (eg. is root span always checked first and does it have this set?)

Reply:
No, the layer only targets span who's name is "substrate-node"